File: Stats.CPP/Basic_QC.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "Basic_QC.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "Basic_Form"
#pragma resource "*.dfm"
TfrmBasicQC *frmBasicQC;
//---------------------------------------------------------------------------
__fastcall TfrmBasicQC::TfrmBasicQC(TComponent* Owner)
	: TfrmBasic(Owner)
{
	Edit1->Text = FormatFloat("0.00",99.8);
	Confidence = 99.8;
	// Setup QC series and it's properties
	Series1 = new TQCSeries(Chart1);
	Series1->SeriesColor = clBlue;
	Series1->ColorEachLine = False;
	Series1->ColorEachPoint = True;
	Series1->LinePen->Width = 2;
	Series1->Pointer->Style = psTriangle;
	Series1->Pointer->VertSize = 3;
	Series1->Pointer->HorizSize = 3;
	Series1->ControlLimitPen->Style = psDash;
	Series1->ControlLimitPen->Color = clRed;
	Series1->CLPen->Color = clBlue;
	Series1->CLPen->Style = psDash;
	Series1->ParentChart = Chart1;
}
//---------------------------------------------------------------------------
void __fastcall TfrmBasicQC::Recalculate()
{
	BeforeRecalculate();
	Chart1->AutoRepaint = false;
	for (int i = 0;i<Chart1->SeriesCount();i++)
		if (dynamic_cast<TQCSeries*>(Chart1->Series[i]) != NULL)
		{
			TQCSeries* ser = dynamic_cast<TQCSeries*>(Chart1->Series[i]);
			for (int j = 0 ;j<Chart1->Series[i]->Count();j++)
					if ((*Chart1->Series[i]->YValues)[j]<ser->UCL && (*Chart1->Series[i]->YValues)[j]>ser->LCL)
						ser->ValueColor[j] = ser->SeriesColor;
					else Chart1->Series[i]->ValueColor[j] = clRed;
		}
	Chart1->AutoRepaint = true;
	Chart1->Repaint();
}

void __fastcall TfrmBasicQC::Button1Click(TObject *Sender)
{
	EditSeries(this,Series1);	
}
//---------------------------------------------------------------------------

void __fastcall TfrmBasicQC::Button2Click(TObject *Sender)
{
	ViewValues(Data,Data->Caption,true,false);
	Recalculate();
}
//---------------------------------------------------------------------------

void __fastcall TfrmBasicQC::FormDestroy(TObject *Sender)
{
  Series1->ParentChart = NULL;
	delete Series1;
}
//---------------------------------------------------------------------------

void __fastcall TfrmBasicQC::Edit1Change(TObject *Sender)
{
	if (Showing && Edit1->Text != "")
	{
		try
		{
			Confidence = StrToFloat(Edit1->Text);
		} catch (...) {
			Confidence = 99.8;
		}
		Recalculate();
	}
}
//---------------------------------------------------------------------------

File: Stats.Delphi/Basic_QC.pas
unit Basic_QC;

interface

uses
  Windows, Messages, SysUtils, Classes, Graphics, Controls, Forms, Dialogs,
  Basic_Form, StdCtrls, TeEngine, Series, StatSeries, TeeProcs, Chart,
  ExtCtrls, MtxVec, Math387, ComCtrls, TeeTools;

type
  TfrmBasicQC = class(TfrmBasic)
    Chart1: TChart;
    Panel2: TPanel;
    Label1: TLabel;
    Edit1: TEdit;
    Button1: TButton;
    Button2: TButton;
    GridBandTool1: TGridBandTool;
    procedure Button1Click(Sender: TObject);
    procedure Button2Click(Sender: TObject);
    procedure FormCreate(Sender: TObject);
    procedure FormDestroy(Sender: TObject);
    procedure Edit1Change(Sender: TObject);
  private
    { Private declarations }
  protected
    { Protected declarations }
    procedure Recalculate; virtual;
  public
    Series1: TQCSeries;
    Data : TMtx;
    DrawVec: TVec;
    DeltaInt : double;
    Confidence: double;
    { Public declarations }
  end;

var
  frmBasicQC: TfrmBasicQC;

implementation

Uses EditChar, StatQCSeriesEdit, MtxVecEdit;
{$R *.DFM}

procedure TfrmBasicQC.Button1Click(Sender: TObject);
begin
  EditSeries(Self,Series1);
end;

procedure TfrmBasicQC.Button2Click(Sender: TObject);
begin
  ViewValues(Data,Data.Caption,True,False);
  Recalculate;
end;

procedure TfrmBasicQC.FormCreate(Sender: TObject);
begin
  inherited;
  Data := TMtx.Create;
  DrawVec := TVec.Create;
  Edit1.Text := FormatFloat('0.00',99.8);
  Confidence := 99.8;
  { Setup QC series and it's properties }
  Series1 := TQCSeries.Create(Chart1);
  Series1.SeriesColor := clBlue;
  Series1.ColorEachLine := False;
  Series1.ColorEachPoint := True;
  Series1.LinePen.Width := 2;
  With Series1.Pointer do
  begin
    Style := psTriangle;
    VertSize := 3;
    HorizSize := 3;
  end;
  With Series1.ControlLimitPen do
  begin
    Style := psDash;
    Color := clRed;
  end;
  With Series1.CLPen do
  begin
    Color := clBlue;
    Style := psDash;
  end;
  Series1.ParentChart := Chart1;
end;

procedure TfrmBasicQC.FormDestroy(Sender: TObject);
begin
  Series1.ParentChart := nil;
  Series1.Free;
  Data.Free;
  DrawVec.Free;
  inherited;
end;

procedure TfrmBasicQC.Recalculate;
var i,j: Integer;
begin
  With Chart1 do
  begin
    AutoRepaint := False;
    for i := 0 to SeriesCount-1 do
      if Series[i] is TQCSeries then
      With (Series[i] as TQCSeries) do
        for j := 0 to Count-1 do
          if (YValues[j]<UCL) and (YValues[j]>LCL) then ValueColor[j] := SeriesColor
          else ValueColor[j] := clRed;
    AutoRepaint := True;
    Repaint;
  end;
end;

procedure TfrmBasicQC.Edit1Change(Sender: TObject);
begin
  if (Showing) and (Edit1.Text <> '') then
  begin
    try
      Confidence := StrToFloat(Edit1.Text);
    except
      Confidence := 99.8;
    end;
    Recalculate;
  end;
end;

end.
